[WIP, ENH] Adds cumulative methods to ea

datajanko opened the pull request:
- [x] closes #28385
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Problem specific

- [X] Add abstract tests similar to pandas/tests/extension/base/reduce.py
- [X] Add TestFixtures for Accumulations
- [x] Implement Accumulation for IntegerArray
- [ ] Implement Accumulation for DecimalArray - out of scope
- [ ] Implement Accumulation for DatetileLikeArrayMixin ? - out of scope
- [x] Use BaseNoAccumulateTest where applicable - implemented this for categorical

datajanko committed:
Merge pull request #1 from pandas-dev/master

updates pandas

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

datajanko committed:
define accumulation interface for ExtensionArrays

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -407,6 +409,7 @@ def isna(self) -> ArrayLike:
 
         * ``na_values._is_boolean`` should be True
         * `na_values` should implement :func:`ExtensionArray._reduce`
+        * `na_values` should implement :func:`ExtensionArray._accumulate`
```
Is this necessarily true?

The previous comment is there because we do something like `Series.array.isna()._reduce('any')` (not sure where).

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -992,6 +995,35 @@ def _ndarray_values(self) -> np.ndarray:
+        ----------
+        name : str
+            Name of the function, supported values are:
+            { cummin, cummax, cumsum, cumprod }.
```
Make this a list.

Can you put the signature of the callable? Do we know what this is?

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -992,6 +995,35 @@ def _ndarray_values(self) -> np.ndarray:
+            Name of the function, supported values are:
+            { cummin, cummax, cumsum, cumprod }.
+        skipna : bool, default True
+            If True, skip NaN values.
```
NaN -> NA.

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -992,6 +995,35 @@ def _ndarray_values(self) -> np.ndarray:
+            If True, skip NaN values.
+        **kwargs
+            Additional keyword arguments passed to the accumulation function.
+            Currently, no is the only supported kwarg.
```
What's "no"?

datajanko replied:
just intended to show my insecurity here, if we need it.

TomAugspurger reviewed (commented):
Can you add base tests similar to the ones in `pandas/tests/extension/base/reduce.py`?

Then you'll need to implement `_accumulate` on

* IntegerArray
* DatetimeLikeArrayMixin? (or maybe a subclass?)
* DecimalArray

What's the return type on `ExtensionArray[T]._accumulate`? Is it `ExtensionArray[T]`? Or can you return an ExtensionArray of a different type?

For example, I could imagine a `BoolArray.cumsum()` that returns an IntegerArray.

datajanko commented:
Sorry, I did not intend to have a proper code review, essentially, I just replaced reduce with accumulate and created duplicates. Will have a deeper dive into the consequences soon.

I'll follow your guidance and start implementing the test soon.

As you pointed out, we should often see a different Type `ExtensionArray[S]`. Additionally to your example, `cummax` could potentially return signed integers. E.g. on an array `[1, -1, 1, -1, ... ]`.

datajanko committed:
reformulate doc string

datajanko committed:
creates baseExtension tests for accumulate

datajanko commented:
I was expecting that in extension arrays, the `_reduce` function is used to define e.g. `sum`, `max` 
etc. However, this does not seem to happen (maybe as of now). So I'm trying to find, where an ExtensionArray gets these functions. And I'm only guessing:

I'd assume that an extension array typically is seen as a Series. Here the `_reduce` function is defined. Moreover, we inherit from `base.IndexOpsMixin` which defines e.g. `max`, `min` etc.

On the other hand, on the Series, we perform e.g. `_add_numeric_operations` (defined for NDFrames (which inherit from pandas objects) which uses the `_reduce` function of the object.

So what is the right interpretation here? Thanks in advance

datajanko committed:
adds fixtures for numeric_accumulations

jbrockmendel commented:
> I was expecting that in extension arrays, the _reduce function is used to define e.g. sum, max
etc. However, this does not seem to happen (maybe as of now). So I'm trying to find, where an ExtensionArray gets these functions.

This is up to EA authors how to implement.  e.g. `sum` is not necessarily well-defined for some EAs (like DatetimeArray) so we can't define it in the general case

datajanko commented:
Thanks for getting back to me. I just realized, that e.g. IntegerArray only uses `_reduce` and does not provide a `sum` property (which I expected). I hope to have a bit more time soon.

datajanko committed:
fixes typos

datajanko committed:
adds accumulate tests for integer arrays

datajanko committed:
fixes typo

alimcmaster1 commented:
@datajanko - is this still active? Mind addressing the comments above and fix up the tests. Feel free to post on here if you run into any issues!

datajanko commented:
Yes, I'm still working on this. I think so far I managed to create the abstract test classes. Next step is to implement the functionality for integer arrays. Which comments do you mean in particular?

datajanko commented:
@TomAugspurger I just realized that your comment on the dtypes is crucial. Doing the cumsum on the first 100 digits, will directly raise this issue for Int8 dtypes. So what should be the strategy here?

In terms of cumsum, an implementation could be:

replace the data at masked positions with 0 (neutral element of addition). Compute the cumsum (using numpy's cumsum). Now find the max and min value. And determine an underlying numpy type that is suitable. Now create a new array (or change the params of the existing array if possible) and return this. 

The same would hold for cumprod (but leaving the values at the masked position to be 1).

cummax and cummin would not have this issue of exceeding the bounds of the dtype

TomAugspurger commented:
Replacing masked values with 0 (or 1 for prod) seems reasonable. We'll need to decide if we want to match NumPy's behavior (which silently overflows) or whether we want to catch that. Either is fine by me.

> Compute the cumsum (using numpy's cumsum). Now find the max and min value.

I don't know that that'll work, because NumPy silently overflows.

datajanko commented:
On the _data of an extension array, cumsum does not flow over silently, but does the correct aggregation
```
np.array([  1,   2,   3,   4,   5,   6,   7,   8,   1,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,   1,  99, 100], dtype='int8').cumsum()
```
gives
```
array([   1,    3,    6,   10,   15,   21,   28,   36,   37,   47,   58,
         70,   83,   97,  112,  128,  145,  163,  182,  202,  223,  245,
        268,  292,  317,  343,  370,  398,  427,  457,  488,  520,  553,
        587,  622,  658,  695,  733,  772,  812,  853,  895,  938,  982,
       1027, 1073, 1120, 1168, 1217, 1267, 1318, 1370, 1423, 1477, 1532,
       1588, 1645, 1703, 1762, 1822, 1883, 1945, 2008, 2072, 2137, 2203,
       2270, 2338, 2407, 2477, 2548, 2620, 2693, 2767, 2842, 2918, 2995,
       3073, 3152, 3232, 3313, 3395, 3478, 3562, 3647, 3733, 3820, 3908,
       3997, 4087, 4178, 4270, 4363, 4457, 4552, 4648, 4745, 4746, 4845,
       4945])
```
So apparently, an easy implementation would be to always return `Int64 Dtype`, which is not optimal but I think a suitable solution at least at the beginning.

datajanko committed:
first implementation of cumsum

datajanko commented:
Okay, this should be a rough idea in the happy cases. Right now, the tests still fail since I haven't connected `._accumulate(name="cumsum")` to the cumsum function of the Series

WillAyd commented:
@datajanko is this still active?

datajanko commented:
Yes in fact it is, but unfortunately time is scarce for me currently. Should be better by the end of February. If you want to allocate more/different resources on this issue, I'm of course fine with that.

datajanko committed:
Merge pull request #2 from pandas-dev/master

Updates fork

WillAyd commented:
@datajanko still active? If so can you fix up and try to get green?

datajanko committed:
Merge pull request #3 from pandas-dev/master

updates from upstream

datajanko committed:
merges master

datajanko committed:
stashed merge conflict

pep8speaks commented:
Hello @datajanko! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`pandas/core/generic.py`](https://github.com/pandas-dev/pandas/blob/73363bfdde30eda50a90b26688f788eb9207de49/pandas/core/generic.py):

> [Line 11201:1](https://github.com/pandas-dev/pandas/blob/73363bfdde30eda50a90b26688f788eb9207de49/pandas/core/generic.py#L11201): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 11229:89](https://github.com/pandas-dev/pandas/blob/73363bfdde30eda50a90b26688f788eb9207de49/pandas/core/generic.py#L11229): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 88 characters)
> [Line 11232:1](https://github.com/pandas-dev/pandas/blob/73363bfdde30eda50a90b26688f788eb9207de49/pandas/core/generic.py#L11232): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace





* In the file [`pandas/tests/extension/test_categorical.py`](https://github.com/pandas-dev/pandas/blob/73363bfdde30eda50a90b26688f788eb9207de49/pandas/tests/extension/test_categorical.py):

> [Line 164:1](https://github.com/pandas-dev/pandas/blob/73363bfdde30eda50a90b26688f788eb9207de49/pandas/tests/extension/test_categorical.py#L164): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

datajanko committed:
fixes formatting

datajanko commented:
yes, I'm working on this. Should soon be able to have some green tests. I realized that my tests were just wrong.

A different related question: I have a series and call e.g. cumsum on the underlying extension array. What is the recommended way to turn this ExtensionArray again into a pd.Series?

jbrockmendel commented:
> What is the recommended way to turn this ExtensionArray again into a pd.Series?

`pd.Series(arr)`

datajanko committed:
first green test for integer extension arrays and cumsum

datajanko committed:
first passing tests for cummin and cummax

datajanko committed:
utilizes na_accum_func

datajanko committed:
removes delegation leftover

datajanko commented:
Hey, I finally seem to habe a working solution (no tests yet, and somehow the pre-commits currently fail on my system) but in a jupyter notebook I seem to get good results for data generate by

```
def make_data():
    return list(range(1, 9)) + [np.nan] + list(range(10, 98)) + [np.nan] + [99, 100]
```

However, these data are not suitable for `cumprod` due to integer overflows. Is there any suggestion on how to test this? Is there any other comment? Is using `na_accum_func` okay?

How do we want to set the further scope of the ticket? Next step for me is finalizing the tests for Integer Arrays. Sorry for the long time, I didn't expect to dive so deep into pandas internals.

datajanko committed:
creates running tests

datajanko committed:
Merge branch 'master' into 28385-add-cumulative-methods-to-EA

datajanko committed:
removes ABCExtensionArray Type hint

datajanko committed:
Merge pull request #4 from pandas-dev/master

merges upstream

datajanko committed:
Merge branch 'master' into 28385-add-cumulative-methods-to-EA

datajanko committed:
removes clutter from generic.py

datajanko committed:
removes clutter in _accumulate

datajanko commented:
so to test the cumulative product, I decided to just take the first 20 rows of the test. A suggestion to organize the tests differently would be welcome. What about the NoAccumulateTests?

TomAugspurger commented:
Haven't looked at the rest yet, but

> About TypeHints, would we want to add from __future__ import annotations to allow a typehint for _accumulate -> ExtensionArray?

Our oldest python doesn't support that. You can have the return value be `-> 'ExtensionArray'` (with the quotes).

datajanko committed:
adds typehints for ExtensionArray and IntegerArray

datajanko commented:
Sorry, I just realized, that I'm not actually calling the `_accumulate`-function in the respective extension arrays but rather put the burden of accumulation in the na_accumulate_function.

My suggestion would now be to do the following:

In ```
        def block_accum_func(blk_values):
            values = blk_values.T if hasattr(blk_values, "T") else blk_values

            result = nanops.na_accum_func(values, accum_func, skipna=skipna)

            result = result.T if hasattr(result, "T") else result
            return result

        result = self._mgr.apply(block_accum_func)
```
in generic.py

I would insert (e.g.)

```
if is instance(blk_values, ExtensionArray):
    result = blk_values._accumulate(...)
else:
    result = nanops.na_accum_func(...)
```

to compute the result. Thanks to the failing tests on the categorical I realized this issue

datajanko committed:
delegates the accumulate calls to extension arrays

datajanko commented:
Okay, so I was able do actually delegate the execution to the extension arrays. 

Open points that remain:
- is the approach okay?
- how shall the scope of the ticket be? Now I know what to do at least :)
- Testing of cumprod seems awkward
- I'm doing some kind of integration test by only looking at series and not on extension arrays directly. Shall this be changed?

datajanko committed:
removes diff in nanops

datajanko committed:
removes unwanted pattern

datajanko committed:
makes output types for sum and prod explicit

datajanko commented:
Any idea on why (except the docs) the tests are failing? Is this a common issue?

datajanko committed:
makes the base accumulate test more general by not comparing types

datajanko committed:
implements accumulation for boolean arrays

datajanko commented:
Concerning boolean arrays: Would it make sense, in principal to implement them as a subclass of integer arrays and just add additional functions specific to boolean arrays? For the accumulation function - at least - we wouldn't need to write any additional code here

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -1023,6 +1025,42 @@ def _concat_same_type(
+        TypeError : subclass does not define accumulations
+        """
+        raise TypeError(
+            "cannot perform {name} with type {dtype}".format(
```
Can use fstrings here.

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -1023,6 +1025,42 @@ def _concat_same_type(
+            - cumprod
+        skipna : bool, default True
+            If True, skip NA values.
+        **kwargs
```
May be for numpy compatibility (axis, dtype, out).

datajanko replied:
What are your expectations adding this? Do you purely want to have it (for this ticket) as accessor, which we'll ignore? If not, I'd guess the impact of axis is None, but dtype might be interesting. I.e for cumsum and integer dtypes, we could also provide the target output type, so not defaulting to (U)Int64. But I don't know if this should be part of this issue

TomAugspurger replied:
The idea would be for something like `np.cumsum(pd.array([1, 2]))` to return an IntegerArray. I'm not sure what all is required for that to work.

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -688,6 +688,35 @@ def cmp_method(self, other):
         name = f"__{op.__name__}"
         return set_function_name(cmp_method, name, cls)
 
+    def _accumulate(
```
I suspect this can be put on `pandas.core.arrays.masked.BaseMaskedArray`.

TomAugspurger replied:
May also want to look at `core.array_algos.masked_reductions` for inspiration. I suspect it'll look somewhat similar.

datajanko replied:
This is a good idea to deduplicate the code for boolean and integer arrays. Thanks for the hint

TomAugspurger replied:
Actually, for deduplication, how about defining this as

```python
from pandas.arrays import IntegerArray
return IntegerArray(self._data.astype(int), self._mask)._accumulate(name, skipna, **kwargs)
```

datajanko replied:
I guess this would upcast a boolean array always to an integer array. This might not be desired for `cummax`, `cummin` and `cumprod`. However, I'd be fine with that approach

datajanko replied:
I chose 'int8' as dtype to keep the arrays as small as possible

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -11191,7 +11192,26 @@ def block_accum_func(blk_values):
+        if isinstance(self.values, ExtensionArray):
+            result = self.values._accumulate(name, skipna, **kwargs)
+        else:
+            result = self._mgr.apply(block_accum_func)
```
I think we always want `self._mgr.apply(block_accum_func)`.

We'll just update `block_accum_func` to handle EAs.  something like

```
if is_extension_array_dtype(self.dtype):
    result = self.array._accumulate(name, skipna, **kwargs)
else:
    ...
```

datajanko replied:
Thanks, I was sensing something like this, but I somehow was not finding the accessor of the array

datajanko committed:
uses f-string in base.py

datajanko committed:
uses blockmanager also for extension arrays

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

datajanko committed:
merges master

datajanko committed:
fixes flake8 issues

jreback commented:
can you merge master and we'll see where we can go with this

datajanko commented:
Yes, will do it this week and probably also some refactoring. Sorry for the delay

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

datajanko committed:
merges master

datajanko commented:
ok, master should be merged. I'd suggest to move `_accumulate` also to masked array as @TomAugspurger 

The, I'd suggest to just push this an not implement the accumulation for other extension arrays, at least for now. What do you think?

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -11524,13 +11524,37 @@ def cum_func(self, axis=None, skipna=True, *args, **kwargs):
             values = blk_values.T if hasattr(blk_values, "T") else blk_values
 
-            result = nanops.na_accum_func(values, accum_func, skipna=skipna)
+            if is_extension_array_dtype(values.dtype):
+                result = values._accumulate(name, skipna, **kwargs)
+            else:
+                result = nanops.na_accum_func(values, accum_func, skipna=skipna)
```
Might want to do this if statement down in `na_accum_func`. So we always call `nanops.na_accum_func(values, accum_func, skipna)` and it's what dispatches based on whether `values` is an EA or not.

datajanko replied:
Yes, we could do this but note the following, in `_accumulate` we explicitly use the name (just like in _reduce, to map infer the accum_func we actually want to use).
This will leave us with two options
1. break similarity with `reduce` and just abandon the `name`. This will imply that in the `_accumulate` signature, we replace `name` by `accum_func` and could get rid of the `**kwargs` as we only have the arguments passed to `na_accum_func`
2. change the API of `na_accum_func`, to include the `name` (as a kwarg?) and further kwargs as well?

@TomAugspurger what would be your favorite solution?

TomAugspurger replied:
Mmm thanks for the explanation. Let's keep it similar to `_reduce` I think.

datajanko replied:
@jreback here is an explanation why I chose to avoid na_accum_func

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -11524,13 +11524,37 @@ def cum_func(self, axis=None, skipna=True, *args, **kwargs):
 
         result = self._mgr.apply(block_accum_func)
 
+        # y = com.values_from_object(self).copy()
```
What's the plan here?

TomAugspurger commented on pandas/tests/extension/base/accumulate.py:
```diff
@@ -0,0 +1,43 @@
+        op_name = all_numeric_accumulations
+        s = pd.Series(data)
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", RuntimeWarning)
```
Why is this needed?

datajanko replied:
I simply copied from reduce.py. Comment there is: `# min/max with empty produce numpy warnings`
I'd guess that `cummax/cummin` will do the same. I'll check this later

TomAugspurger commented on pandas/tests/extension/base/accumulate.py:
```diff
@@ -0,0 +1,43 @@
+    make sense for numeric/boolean operations.
+    """
+
+    def check_accumulate(self, s, op_name, skipna):
+        result = getattr(s, op_name)(skipna=skipna)
+        expected = getattr(s.astype("float64"), op_name)(skipna=skipna)
+        tm.assert_almost_equal(result, expected, check_dtype=False)
```
I believe this is copied from the other tests, but what do people think about structuring things a bit differently? This is doing a lot of stuff, making it hard to customize for a specific type.

So I'd recommend something like

```python
def get_expected(self, s, op_name, skipna):
    return getattr(s.astype("float64"), op_name)(skipna=skipna)
```

and then the test can do the `self.assert_series_equal(result, expected)`

TomAugspurger reviewed (commented):
I'm OK with just doing this for Integer & Boolean for now.

datajanko committed:
removes uncommented code

datajanko committed:
adds todo for runtime warning

datajanko committed:
reuses integer array to accumulate for booleans

datajanko commented:
On pandas-dev I'm seeing failing tests for date time arrays and cummin/cummax. My feeling is that having at least these two functions there is necessary. Raising a not implemented error would be strange

datajanko committed:
removes runtimewarning catching

jbrockmendel commented:
> On pandas-dev I'm seeing failing tests for date time arrays and cummin/cummax. My feeling is that having at least these two functions there is necessary. Raising a not implemented error would be strange

Can you post a representative traceback?

datajanko commented:
sure:

<details> <summary>https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=37916&view=logs&j=a67b4c4c-cd2e-5e3c-a361-de73ac9c05f9&t=33d2fdd0-c376-5f94-e6d3-957bdd23a3b8&l=134</summary>
```

        The underlying data type might change
        # TODO Clarify
    
        Parameters
        ----------
        name : str
            Name of the function, supported values are:
            # TODO Add function signatures
            - cummin
            - cummax
            - cumsum
            - cumprod
        skipna : bool, default True
            If True, skip NA values.
        **kwargs
            Additional keyword arguments passed to the accumulation function.
            # TODO check if kwargs are needed
            Currently, there is no supported kwarg.
    
        Returns
        -------
        array
    
        Raises
        ------
        TypeError : subclass does not define accumulations
        """
>       raise TypeError(f"cannot perform {name} with type {self.dtype}")
E       TypeError: cannot perform cummax with type datetime64[ns, US/Pacific]

pandas/core/arrays/base.py:1121: TypeError
```

</details>

datajanko committed:
removes TODOs

datajanko committed:
adds accumulate to autosummary

datajanko commented:
I can reproduce these errors locally

datajanko commented:
Apparently,

in
```
def block_accum_func(blk_values):
            values = blk_values.T if hasattr(blk_values, "T") else blk_values

            if is_extension_array_dtype(values.dtype):
                result = values._accumulate(name, skipna, **kwargs)
            else:
                result = nanops.na_accum_func(values, accum_func, skipna=skipna)

            result = result.T if hasattr(result, "T") else result
            return result
```
the predicate `is_extension_array_dtype(values.dtype)` evaluates to true if a timezone is provided and to false if this is not the case.

What would be the desired way of handling this here? I guess we shouldn't expect our users to localize timestamps to be able to use the two cumulative operations.

A hack would be to only check for masked arrays. But I guess this is not what we want to do. Any advice?

jbrockmendel commented:
> is_extension_array_dtype(values.dtype)

In most of the relevant places we check for this along with `or dtype.kind in ["m", "M"]` to include DatetimeArray/TimedeltaArray

datajanko committed:
excludes datetime from propagating to _accumulate

datajanko commented:
hmm, any idea on why the tests keep failing? I don't even see which tests fail, but it seems related to an import error. Could someone trigger the pipeline again?

datajanko committed:
uses pandas.testing instead of pandas.util.testing in accumulate

datajanko committed:
replaces assert_almost_equal with assert_series_equal

datajanko committed:
dtypes to lowercase

datajanko committed:
lowercase of uint and int64 dtype in _accumulate

datajanko commented:
Hey, why is ```##[error]pandas/tests/extension/base/accumulate.py:18:        tm.assert_series_equal(result, expected, check_dtype=False)``` an unwanted pattern and how to resolve this?

simonjayhawkins commented:
> Hey, why is `##[error]pandas/tests/extension/base/accumulate.py:18: tm.assert_series_equal(result, expected, check_dtype=False)` an unwanted pattern and how to resolve this. Any suggestions?

use `self.assert_series_equal` since some assertion helpers are overridden for certain EA types.

datajanko committed:
uses hint of @simonjayhawkins concerning assert series equals

datajanko committed:
Merge branch 'master' into 28385-add-cumulative-methods-to-EA

datajanko commented:
I'm finding a different AssertionError now
```
=================================== FAILURES ===================================
1409______________ test_chunks_have_consistent_numerical_type[python] ______________
1410[gw0] linux -- Python 3.7.7 /home/travis/miniconda3/envs/pandas-dev/bin/python
1411
1412all_parsers = <pandas.tests.io.parser.conftest.PythonParser object at 0x7fd72283a350>
1413
1414    def test_chunks_have_consistent_numerical_type(all_parsers):
1415        parser = all_parsers
1416        integers = [str(i) for i in range(499999)]
1417        data = "a\n" + "\n".join(integers + ["1.0", "2.0"] + integers)
1418    
1419        # Coercions should work without warnings.
1420        with tm.assert_produces_warning(None):
1421>           result = parser.read_csv(StringIO(data))
1422
1423pandas/tests/io/parser/test_common.py:1148: 
1424_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
1425
1426self = <contextlib._GeneratorContextManager object at 0x7fd6c5c3a510>
1427type = None, value = None, traceback = None
1428
1429    def __exit__(self, type, value, traceback):
1430        if type is None:
1431            try:
1432>               next(self.gen)
1433E               AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning("unclosed <ssl.SSLSocket fd=8, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('10.20.255.23', 56320), raddr=('52.216.29.108', 443)>"), '/home/travis/build/pandas-dev/pandas/pandas/io/parsers.py', 3004), ('ResourceWarning', ResourceWarning("unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('10.20.255.23', 56322), raddr=('52.216.29.108', 443)>"), '/home/travis/build/pandas-dev/pandas/pandas/io/parsers.py', 3004)]
1434
1435../../../miniconda3/envs/pandas-dev/lib/python3.7/contextlib.py:119: AssertionError
```

Any Ideas what this is?

jbrockmendel commented:
> Any Ideas what this is?

This looks unrelated, generally happens when some other test is leaving behind an open file or network connection.  If it is only happening intermittently, you can ignore it for this PR (but if you want to track it down for a separate PR that'd be great!)

datajanko commented:
Should I create an issue for that?

datajanko committed:
adds whatsnew entry

datajanko commented:
any idea how an added whatsnew entry leads to more than 800 failing tests that seem to be completely unrelated?

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

datajanko committed:
Merge branch 'master' into 28385-add-cumulative-methods-to-EA

datajanko commented:
good news, the specific travis error is gone. Now we only see unrelated errors that happen as well on other PRs. How can we proceed here @jbrockmendel? I'd really like to get this merged before Friday

jreback reviewed (changes requested):
i'll look soon

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into master

datajanko committed:
Merge branch 'master' into 28385-add-cumulative-methods-to-EA

datajanko committed:
moves changes to 1.2.0

datajanko commented:
I guess that issue is also unrelated?

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into master

datajanko committed:
Merge branch 'master' into 28385-add-cumulative-methods-to-EA

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -267,6 +267,7 @@ Sparse
 ExtensionArray
 ^^^^^^^^^^^^^^
 
+- Added :meth:`api.extensionExtensionArray._accumulate` to the extension array interface. Implements this interface for :class: `IntegerArray` and :class: `BooleanArray` such that type coercion to `object` is avoided (:issue:`28385`)
```
can you instead frame this as what a user would understand as `_accumulate` is a private method

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -1120,7 +1122,37 @@ def _concat_same_type(
     # of objects
     _can_hold_na = True
 
+    def _accumulate(self, name, skipna=True, **kwargs) -> "ExtensionArray":
```
can you type name, skipna

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -553,6 +553,38 @@ def cmp_method(self, other):
+            "cumsum": (np.cumsum, 0),
+            "cummin": (np.minimum.accumulate, data.max()),
+        }[name]
+        from ..nanops import _get_values
```
use absolute imports

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -553,6 +553,38 @@ def cmp_method(self, other):
+            "cummax": (np.maximum.accumulate, data.min()),
+            "cumsum": (np.cumsum, 0),
+            "cummin": (np.minimum.accumulate, data.max()),
+        }[name]
```
this would raise KeyError on an invalid function, but that is not a friendly message to the user, you can either catch it or (do an in on the dict first) and provide a nice ValueError message

datajanko replied:
done

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -553,6 +553,38 @@ def cmp_method(self, other):
+            else:
+                dtype_out = "int64"
+
+        vals = cum_function(values, dtype=dtype_out)
```
you should be able to use nanops.nan_accum_func no?

datajanko replied:
I could use this, but see my comment on https://github.com/pandas-dev/pandas/pull/28509#discussion_r440436191
I'm happy to go in any direction here

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -553,6 +553,38 @@ def cmp_method(self, other):
+
+        # makes target dtypes explicit since CI showed optimal UInt32
+        # dtype on test data occasionally. This was different across systems
+        dtype_out = dtype
```
the nanroutines already handle this

jreback commented on pandas/core/generic.py:
```diff
@@ -11601,7 +11601,13 @@ def cum_func(self, axis=None, skipna=True, *args, **kwargs):
             values = blk_values.T if hasattr(blk_values, "T") else blk_values
 
-            result = nanops.na_accum_func(values, accum_func, skipna=skipna)
+            if is_extension_array_dtype(values.dtype) and values.dtype.kind not in [
```
I would rather have the actual logic live here than splitting it between the array functions and here.

jreback commented on pandas/tests/extension/test_integer.py:
```diff
@@ -249,6 +249,51 @@ class TestBooleanReduce(base.BaseBooleanReduceTests):
     pass
 
 
+class TestNumericAccumulation(base.BaseNumericAccumulateTests):
```
we already have *some* tests for sparse, pls add the AccumulateTests mixin there as well.

jreback commented on pandas/tests/extension/test_integer.py:
```diff
@@ -249,6 +249,51 @@ class TestBooleanReduce(base.BaseBooleanReduceTests):
+                )
+            )
+            tm.assert_series_equal(result, expected)
+        elif op_name in ["cummax", "cummin"]:
```
please try to make this simpler (the entire function), you are repeating lots of things

jreback commented on pandas/tests/extension/test_integer.py:
```diff
@@ -249,6 +249,51 @@ class TestBooleanReduce(base.BaseBooleanReduceTests):
+            )
+            tm.assert_series_equal(result, expected)
+
+        else:
```
this needs a separate test for invalid names

datajanko commented:
@jreback thanks for the review, will take care of your comments soon. Concerning nanops: I guess I had a discussion with @TomAugspurger about this here: https://github.com/pandas-dev/pandas/pull/28509#discussion_r440436191
What is your thought on that?

jbrockmendel commented:
@datajanko can you rebase and ill take a look

datajanko commented:
sorry, time was unexpectedly scarce for me. Things will look better after Monday. I didn't really tackle @jreback's comments.

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into master

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into master

datajanko committed:
merges master

datajanko committed:
merges master

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into master

datajanko committed:
uses na_accum_func

datajanko commented:
@jreback could you elaborate a bit on your comment here: https://github.com/pandas-dev/pandas/pull/28509/files/b33a5df5a042e46a9d28db9a6f4dd37b7bb67ed8#r468754088
which is outdated right now. I currently found some time but I guess this is where I'm struggling:

- I want to define some logic in the extension arrays `_accumulate` function because I eventually might want to run an accumulation on an extension array and not a series, right?
- I need to make some adjustments in generic.py as otherwise, calling cumsum on a series will not trigger the extension array accumulation
- I want to make adjustments in nanops, as we typically delegate there and we want to adjust the fill value. This adjustment needs to take place somewhere.

datajanko commented:
I compared more against `reductions`.
in `generic.py` the `NDFrame` has the `_min_count_stats_function` (and other `stats_functions`) which eventually delegates to `_reduce` (which NDFrame does not have(!)). This happens for example for `sum`.

For accumulations, `NDFrame` provides an `_accum_func` which directly uses `nanops.na_accum_func`, there is no delegation yet to an `_accumulate` function. I fear that to solve this issue holistically, we'd need to something similar as well.

WDYT: @jbrockmendel @TomAugspurger ?

jbrockmendel commented:
>  which NDFrame does not have

If you'd like to add the method in the base class, go for it

> I want to make adjustments in nanops,

i think we're trying to avoid having nanops handling EA dtypes.  Does that affect what you have in mind?

datajanko commented:
Okay, maybe I misunderstood the nanops thing from the previous conversations. E.g see here: https://github.com/pandas-dev/pandas/pull/28509#discussion_r468753639
but we can't use `nanops.nan_accum_func` since we'd want to use different `fill_values` that are data dependent

I'll investigate the situation with `NDFrame`

jreback commented:
if you would like to continue, pls merge master and can have a look

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -532,6 +532,7 @@ ExtensionArray
 - Fixed bug where ``astype()`` with equal dtype and ``copy=False`` would return a new object (:issue:`284881`)
 - Fixed bug when applying a NumPy ufunc with multiple outputs to a :class:`pandas.arrays.IntegerArray` returning None (:issue:`36913`)
 - Fixed an inconsistency in :class:`PeriodArray`'s ``__init__`` signature to those of :class:`DatetimeArray` and :class:`TimedeltaArray` (:issue:`37289`)
+- Added :meth:``api.extensionExtensionArray._accumulate`` to the extension array interface. Implements this interface for :class: `IntegerArray` and :class: `BooleanArray` such that type coercion to `object` is avoided (:issue:`28385`)
```
use single backticks around the `:meth:`

simonjayhawkins replied:
and move to 1.3

jreback commented on pandas/core/nanops.py:
```diff
@@ -1690,6 +1690,28 @@ def na_accum_func(values: ArrayLike, accum_func, skipna: bool) -> ArrayLike:
             result = type(values)._simple_new(  # type: ignore[attr-defined]
                 result, dtype=orig_dtype
             )
+    from pandas.core.arrays import IntegerArray
```
these should be done in NumericArray rather than here

datajanko commented:
okay, I will make time next week. sorry for the long delay. Thanks for your comments. Didn't expect that adding this was so involved in the depths of pandas :)

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

datajanko committed:
merges master

simonjayhawkins commented:
@datajanko it appears that the release note got lost in the merge.

datajanko committed:
delegate to EAs _accumulate function in block mgr

datajanko commented:
@simonjayhawkins 

I'll rewrite the comment fully. It was suggested to not include change notes concerning private functions

Moreover, I guess I'm now seeing a way to not change the nanops module but rather use a masked_accumulations module, similar to masked_reductions. Also delegating to the extension array in

```
        def block_accum_func(blk_values):
            values = blk_values.T if hasattr(blk_values, "T") else blk_values

            if isinstance(values, ExtensionArray):
                result = values._accumulate(name, skipna=skipna, **kwargs)
            else:
                result = nanops.na_accum_func(values, func, skipna=skipna)

            result = result.T if hasattr(result, "T") else result
            return result
```

seems to be the a way to circumvent adding an `_accumulate` interface to series and data frames

datajanko committed:
moves implementation from nanops to masked_accumulations

datajanko committed:
fixes typing annotations in base and masked

datajanko committed:
Merge branch 'master' into 28385-add-cumulative-methods-to-EA

datajanko committed:
fixes merge error

datajanko committed:
fills na values without nanops

datajanko committed:
fixes incorrect call to cumsum and changes to cumprod

datajanko committed:
add _accumulate to boolean

datajanko committed:
makes tests a lot easier - cumprod tests still fail

datajanko commented:
@simonjayhawkins @jreback 

Currently, tests for float masked arrays and test for sparse masked arrays are missing.
Other than that:

I realized that in boolean arrays, the `astype` method does currently not cast to integer arrays. However, in the `_accumulate` function for boolean arrays, I'm somehow using this cast. Do you want me to update the implementation of as type here?

Moreover, cumprod is nasty to test due to overflow errors with the given data. Do you have a suggestion on what to do here?
In a previous implementation I just restricted the data (whenever the op was cumprod) to the first 20 entries.

Is the approach okay, or would you rather prefer a different design somewhere?

datajanko committed:
adds BaseNumericAccumulation for floating masked array

datajanko committed:
tests no numeric accumulations according to _accumulate interface

datajanko commented:
How do you want to scope the ticket?

- Decimal Array is not there anymore it seems.
- It would make sense to add accumulative functions for timedeltas
- It could make sense to add cummin and cummax for timestamps
- I think it wouldn't make sense to add cumulative methods to periods

A `cumsum` method is defined for sparse arrays, but not via the `_accumulate` interface. I didn't see any tests. Do we want to have cumulative functions for sparse arrays?

Currently, we see (just using the standard base tests)
- `cumprod` tests failing for integer apparently due to some overflow issues.
- various cumulative functions failing for floating arrays (all have skipna=True)

Moreover, timestamps and timedeltas seem to implement some cumulative methods via series and specifically:
we see two failing tests

```
FAILED pandas/tests/series/test_cumulative.py::TestSeriesCumulativeOps::test_cummin_datetime64[US/Pacific]
FAILED pandas/tests/series/test_cumulative.py::TestSeriesCumulativeOps::test_cummax_datetime64[US/Pacific]
```

Any comments/suggestions on that?

jbrockmendel commented:
> I think it wouldn't make sense to add cumulative methods to periods

Why not cummin/cummax?

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -1232,6 +1234,37 @@ def _concat_same_type(
+        ------
+        TypeError : subclass does not define accumulations
+        """
+        raise TypeError(f"cannot perform {name} with type {self.dtype}")
```
NotImplementedError?

jbrockmendel commented:
> test_cummin_datetime64

you'll need to implement _accumulate on core.arrays.datetimelike.DatetimeLikeArrayMixin

Be careful of NaT handling, IIRC this took some wrangling, see nanops.na_accum_func

datajanko commented:
> > I think it wouldn't make sense to add cumulative methods to periods
> 
> Why not cummin/cummax?

You're right

datajanko committed:
uses NotImplementedError in base accumulate function

datajanko commented:
Hey, apparently, in my local branch I made some progress when implementing `_accumulate` for date time like arrays. However, whenever I use a series or data frame, I'm failing to call the correct function. In fact, in the [block_accum_func]( https://github.com/pandas-dev/pandas/pull/28509#issuecomment-761749171) I'm only entering the nanops branch of the if-else statement.

What is the desired approach here? 

I was trying to track down what happens a bit further and found this TODO
```
  def _split_op_result(self, result) -> List[Block]:
        # See also: split_and_operate
        if is_extension_array_dtype(result) and result.ndim > 1:
            # TODO(EA2D): unnecessary with 2D EAs
            # if we get a 2D ExtensionArray, we need to split it into 1D pieces
```
It feels, that this is related or am I wrong here?

jbrockmendel commented:
you can ignore the `# TODO(EA2D)`, that is just trying to quantify the amount of complexity we could avoid if/when we have 2D EAs.

inside block_accum_func, before doing the `isinstance(values, ExtensionArray)` check, consider calling `construction.ensure_wrapped_if_datetimelike`

datajanko commented:
Awesome, thanks for the hint. I will have a look at this.

datajanko commented:
Hey, even though I didn't push, I guess I made some progress and have a somewhat working implementation of the arrays.

However, I observed an inconsistency when handling skipna:

```
pd.Series([0.0, None, 1.0]).cummax(skipna=False)
```

gives
```
0    0.0
1    NaN
2    NaN
dtype: float64
```

but

```
pd.Series( pd.array([pd.Timestamp("2020-01-01"), pd.NaT, pd.Timestamp('2020-01-02')])).cummax(skipna=False)
```

gives

```
0   2020-01-01
1   2020-01-01
2   2020-01-02
dtype: datetime64[ns]
```

a different behavior.

Apparently, it would be preferable to have uniform behavior (probably the behavior of the floats). But changing this, would also mean a breaking change. I guess, that we want to avoid this. Correct?

jreback commented on pandas/core/arrays/masked.py:
```diff
@@ -427,3 +427,19 @@ def _reduce(self, name: str, *, skipna: bool = True, **kwargs):
+            op = getattr(masked_accumulations, name)
+            data, mask = op(data, mask, skipna=skipna, **kwargs)
+
+        return type(self)(data, mask, copy=False)
```
this should be indented no?

datajanko replied:
yep, I fixed this locally already.

datajanko committed:
ensures the fill values are data independent

additionally, remove min_count as irrellevant

datajanko committed:
adds accumulation for datetimelikes

in generic.py ensure that datetimelikes are wrapped
create a twin of masked_accumulations for datetimelikes
timedeltas also allow cumsum and cumprod, theoretically

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

datajanko committed:
fixes merge conflicts

datajanko committed:
actually ads datetimelike accumulation algos

datajanko committed:
fixes absolute imports

datajanko committed:
changes error to catch to adhere to changed implementation

datajanko commented:
Hey, right now we see the following, where I'd like to request your feedback @jbrockmendel @jreback 

- cumprod fails for skin=False for float32 and integer arrays. Apparently this is due to buffer overruns. We could restrict the dataset to be a bit smaller. Or do you have another suggestion/opinion on what to do?
- we see `pandas/tests/series/test_cumulative.py` as described in my previous comment. What do we want to do here?
- in `pandas/tests/extension` there is no `test_timedelta.py`. I'm wondering if we want to have accumulation tests also here for datetimelikes?

Apart from that, and except for the typing errors, I guess we are almost done.

jbrockmendel commented:
> in pandas/tests/extension there is no test_timedelta.py. I'm wondering if we want to have accumulation tests also here for datetimelikes?

this can either go in a new extensions/test_datetimelike.py file or in arrays/test_datetimelike.yp

> cumprod fails for skin=False for float32 and integer arrays. Apparently this is due to buffer overruns. We could restrict the dataset to be a bit smaller. Or do you have another suggestion/opinion on what to do?

@jorisvandenbossche ?

jreback commented:
closing as stale, if you want to continue working, please ping.
